drivers/adc/adc_emul.c

```diff
@@ -0,0 +1,546 @@
+	/** Semaphore used to control acquisiton thread */
+	struct k_sem sem;
+
+	/** Stack for acquisition thread */
+	K_KERNEL_STACK_MEMBER(stack,
+			CONFIG_ADC_EMUL_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+int adc_emul_const_value_set(const struct device *dev, unsigned int chan,
+			     uint32_t value)
+{
+	const struct adc_emul_config *config = dev->config;
+	struct adc_emul_data *data = dev->data;
+	struct adc_emul_chan_cfg *chan_cfg;
+
+	if (chan >= config->num_channels) {
+		LOG_ERR("unsupported channel %d", chan);
+		return -EINVAL;
+	}
+
+	chan_cfg = &data->chan_cfg[chan];
+	chan_cfg->input = ADC_EMUL_CONST_VALUE;
+	chan_cfg->const_value = value;
+
+	return 0;
+}
+
+int adc_emul_value_func_set(const struct device *dev, unsigned int chan,
+			    adc_emul_value_func func, void *func_data)
+{
+	const struct adc_emul_config *config = dev->config;
+	struct adc_emul_data *data = dev->data;
+	struct adc_emul_chan_cfg *chan_cfg;
+
+	if (chan >= config->num_channels) {
+		LOG_ERR("unsupported channel %d", chan);
+		return -EINVAL;
+	}
+
+	chan_cfg = &data->chan_cfg[chan];
+	chan_cfg->func = func;
+	chan_cfg->func_data = func_data;
+	chan_cfg->input = ADC_EMUL_CUSTOM_FUNC;
+
+	return 0;
+}
+
+int adc_emul_ref_voltage_set(const struct device *dev, enum adc_reference ref,
+			     uint16_t value)
+{
+	struct adc_driver_api *api = (struct adc_driver_api *)dev->api;
+	struct adc_emul_data *data = dev->data;
+
+	switch (ref) {
+	case ADC_REF_VDD_1:
+		data->ref_vdd = value;
+		break;
+	case ADC_REF_INTERNAL:
+		data->ref_int = value;
+		api->ref_internal = value;
+		break;
+	case ADC_REF_EXTERNAL0:
+		data->ref_ext0 = value;
+		break;
+	case ADC_REF_EXTERNAL1:
+		data->ref_ext1 = value;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Convert @p ref to reference voltage value in mV
+ *
+ * @param data Internal data of ADC emulator
+ * @param ref Select which reference source should be used
+ *
+ * @return Reference voltage in mV
+ * @return 0 on error
+ */
+static uint16_t adc_emul_get_ref_voltage(struct adc_emul_data *data,
+					 enum adc_reference ref)
+{
+	switch (ref) {
+	case ADC_REF_VDD_1:
+		return data->ref_vdd;
+	case ADC_REF_VDD_1_2:
+		return data->ref_vdd / 2;
+	case ADC_REF_VDD_1_3:
+		return data->ref_vdd / 3;
+	case ADC_REF_VDD_1_4:
+		return data->ref_vdd / 4;
+	case ADC_REF_INTERNAL:
+		return data->ref_int;
+	case ADC_REF_EXTERNAL0:
+		return data->ref_ext0;
+	case ADC_REF_EXTERNAL1:
+		return data->ref_ext1;
+	default:
+		return 0;
+	}
+}
+
+static int adc_emul_channel_setup(const struct device *dev,
+				  const struct adc_channel_cfg *channel_cfg)
+{
+	const struct adc_emul_config *config = dev->config;
+	struct adc_emul_chan_cfg *emul_chan_cfg;
+	struct adc_emul_data *data = dev->data;
+
+	if (channel_cfg->channel_id >= config->num_channels) {
+		LOG_ERR("unsupported channel id '%d'", channel_cfg->channel_id);
+		return -ENOTSUP;
+	}
+
+	emul_chan_cfg = &data->chan_cfg[channel_cfg->channel_id];
+
```

Comment:
can you drop this blank line and (perhaps?) the one below? I think it can go a bit far

Reply:
I have moved whole `emul_chan_cfg` block after "ifs" that are checking if provided setup is supported. Channel will not be changed in that case.